File: app/api/webhooks.py
"""
Webhooks API

Webhook management for event notifications:
- Subscribe to VMS events
- Configure delivery endpoints
- Retry policies
- Webhook history and logs
"""
from flask import Blueprint, request, jsonify
from bson import ObjectId
from datetime import datetime, timedelta, timezone
import hashlib
import hmac
import secrets
import json

from app.db import get_db
from app.auth import require_auth, require_company_access
from app.utils import get_current_utc

webhooks_bp = Blueprint('webhooks', __name__)


# Available webhook events
WEBHOOK_EVENTS = [
    'visitor.registered',
    'visitor.updated',
    'visitor.deleted',
    'visitor.blacklisted',
    'visit.scheduled',
    'visit.checked_in',
    'visit.checked_out',
    'visit.cancelled',
    'approval.requested',
    'approval.approved',
    'approval.rejected',
    'evacuation.triggered',
    'evacuation.ended',
    'security.blacklist_match',
    'security.alert'
]


def convert_objectids(obj):
    """Recursively convert ObjectIds to strings"""
    if isinstance(obj, dict):
        return {k: convert_objectids(v) for k, v in obj.items()}
    elif isinstance(obj, list):
        return [convert_objectids(i) for i in obj]
    elif isinstance(obj, ObjectId):
        return str(obj)
    elif isinstance(obj, datetime):
        return obj.isoformat()
    return obj


def generate_webhook_secret():
    """Generate a secure webhook secret"""
    return secrets.token_hex(32)


def sign_payload(payload: dict, secret: str) -> str:
    """Create HMAC signature for webhook payload"""
    payload_bytes = json.dumps(payload, sort_keys=True, default=str).encode('utf-8')
    signature = hmac.new(secret.encode('utf-8'), payload_bytes, hashlib.sha256).hexdigest()
    return f"sha256={signature}"


@webhooks_bp.route('/events', methods=['GET'])
@require_company_access
def list_available_events():
    """List all available webhook events"""
    return jsonify({
        'events': WEBHOOK_EVENTS,
        'count': len(WEBHOOK_EVENTS)
    }), 200


@webhooks_bp.route('/subscriptions', methods=['GET'])
@require_company_access
def list_subscriptions():
    """List all webhook subscriptions for a company."""
    try:
        company_id = request.args.get('companyId') or getattr(request, 'company_id', None)
        
        if not company_id:
            return jsonify({'error': 'Company ID is required'}), 400
        
        db = get_db()
        webhooks = db['webhooks']
        
        subs = list(webhooks.find({'companyId': company_id}))
        
        for sub in subs:
            sub.pop('secret', None)
        
        return jsonify({
            'subscriptions': convert_objectids(subs),
            'count': len(subs)
        }), 200
        
    except Exception as e:
        print(f"Error listing webhooks: {e}")
        return jsonify({'error': str(e)}), 500


@webhooks_bp.route('/subscriptions', methods=['POST'])
@require_company_access
def create_subscription():
    """Create a new webhook subscription."""
    try:
        data = request.json or {}
        company_id = data.get('companyId') or getattr(request, 'company_id', None)
        url = data.get('url')
        events = data.get('events', [])
        
        if not company_id:
            return jsonify({'error': 'Company ID is required'}), 400
        if not url:
            return jsonify({'error': 'URL is required'}), 400
        if not events:
            return jsonify({'error': 'At least one event is required'}), 400
        
        invalid_events = [e for e in events if e not in WEBHOOK_EVENTS and e != '*']
        if invalid_events:
            return jsonify({'error': f'Invalid events: {invalid_events}'}), 400
        
        secret = generate_webhook_secret()
        
        db = get_db()
        webhooks = db['webhooks']
        
        webhook_doc = {
            '_id': ObjectId(),
            'companyId': company_id,
            'name': data.get('name', f'Webhook for {url}'),
            'url': url,
            'events': events,
            'secret': secret,
            'headers': data.get('headers', {}),
            'retryPolicy': data.get('retryPolicy', {
                'maxRetries': 3,
                'retryDelaySeconds': [60, 300, 900]
            }),
            'active': True,
            'createdAt': get_current_utc(),
            'createdBy': getattr(request, 'user_id', 'system'),
            'lastDelivery': None,
            'deliveryCount': 0,
            'failureCount': 0
        }
        
        webhooks.insert_one(webhook_doc)
        
        return jsonify({
            'message': 'Webhook subscription created',
            'subscriptionId': str(webhook_doc['_id']),
            'secret': secret,
            'events': events,
            'note': 'Store the secret securely. It will not be shown again.'
        }), 201
        
    except Exception as e:
        print(f"Error creating webhook: {e}")
        import traceback
        traceback.print_exc()
        return jsonify({'error': str(e)}), 500


@webhooks_bp.route('/subscriptions/<subscription_id>', methods=['DELETE'])
@require_company_access
def delete_subscription(subscription_id):
    """Delete a webhook subscription"""
    try:
        db = get_db()
        webhooks = db['webhooks']
        
        result = webhooks.delete_one({'_id': ObjectId(subscription_id)})
        
        if result.deleted_count == 0:
            return jsonify({'error': 'Subscription not found'}), 404
        
        return jsonify({'message': 'Subscription deleted'}), 200
        
    except Exception as e:
        print(f"Error deleting webhook: {e}")
        return jsonify({'error': str(e)}), 500


@webhooks_bp.route('/subscriptions/<subscription_id>/test', methods=['POST'])
@require_company_access
def test_webhook(subscription_id):
    """Send a test webhook delivery."""
    try:
        import requests as http_requests
        
        db = get_db()
        webhooks = db['webhooks']
        
        webhook = webhooks.find_one({'_id': ObjectId(subscription_id)})
        if not webhook:
            return jsonify({'error': 'Subscription not found'}), 404
        
        test_payload = {
            'event': 'webhook.test',
            'timestamp': get_current_utc().isoformat(),
            'data': {
                'message': 'This is a test webhook delivery',
                'subscriptionId': subscription_id
            }
        }
        
        signature = sign_payload(test_payload, webhook['secret'])
        
        headers = {
            'Content-Type': 'application/json',
            'X-Webhook-Signature': signature,
            'X-Webhook-Event': 'webhook.test',
            'X-Webhook-Delivery-Id': str(ObjectId())
        }
        headers.update(webhook.get('headers', {}))
        
        try:
            response = http_requests.post(
                webhook['url'],
                json=test_payload,
                headers=headers,
                timeout=10
            )
            
            success = 200 <= response.status_code < 300
            
            return jsonify({
                'success': success,
                'statusCode': response.status_code,
                'message': 'Test delivery successful' if success else 'Test delivery failed'
            }), 200
            
        except Exception as delivery_error:
            return jsonify({
                'success': False,
                'error': str(delivery_error),
                'message': 'Failed to connect to webhook URL'
            }), 200
        
    except Exception as e:
        print(f"Error testing webhook: {e}")
        return jsonify({'error': str(e)}), 500


@webhooks_bp.route('/deliveries', methods=['GET'])
@require_company_access
def get_delivery_history():
    """Get webhook delivery history."""
    try:
        company_id = request.args.get('companyId') or getattr(request, 'company_id', None)
        subscription_id = request.args.get('subscriptionId')
        limit = int(request.args.get('limit', 50))
        
        if not company_id:
            return jsonify({'error': 'Company ID is required'}), 400
        
        db = get_db()
        deliveries = db['webhook_deliveries']
        
        query = {'companyId': company_id}
        if subscription_id:
            query['subscriptionId'] = ObjectId(subscription_id)
        
        history = list(deliveries.find(query).sort('timestamp', -1).limit(limit))
        
        return jsonify({
            'deliveries': convert_objectids(history),
            'count': len(history)
        }), 200
        
    except Exception as e:
        print(f"Error getting delivery history: {e}")
        return jsonify({'error': str(e)}), 500

File: server/routes/webhooks.js
/**
 * Webhooks API
 * Webhook management for integrations
 * Matching Python app/api/webhooks.py
 */
const express = require('express');
const router = express.Router();
const { ObjectId } = require('mongodb');

const { getDb } = require('../db');
const { requireCompanyAccess } = require('../middleware/auth');
const { convertObjectIds, isValidObjectId, validateRequiredFields } = require('../utils/helpers');

/**
 * GET /api/webhooks
 * List all webhooks for a company
 */
router.get('/', requireCompanyAccess, async (req, res, next) => {
    try {
        const companyId = req.query.companyId;

        if (!companyId) {
            return res.status(400).json({ error: 'Company ID is required.' });
        }

        const db = getDb();
        let query;
        if (isValidObjectId(companyId)) {
            query = { companyId: new ObjectId(companyId) };
        } else {
            query = { companyId };
        }

        const webhooks = await db.collection('webhooks').find(query).toArray();

        res.json({ webhooks: convertObjectIds(webhooks) });
    } catch (error) {
        console.error('Error listing webhooks:', error);
        next(error);
    }
});

/**
 * POST /api/webhooks
 * Create a new webhook
 */
router.post('/', requireCompanyAccess, async (req, res, next) => {
    try {
        const data = req.body;

        const validation = validateRequiredFields(data, ['companyId', 'url', 'events']);
        if (!validation.valid) {
            return res.status(400).json({ error: `Missing required fields: ${validation.missing.join(', ')}` });
        }

        // Validate URL
        try {
            new URL(data.url);
        } catch {
            return res.status(400).json({ error: 'Invalid webhook URL' });
        }

        // Validate events
        const validEvents = [
            'visitor.registered', 'visitor.updated', 'visitor.deleted',
            'visit.scheduled', 'visit.checked_in', 'visit.checked_out', 'visit.cancelled',
            'employee.registered', 'employee.updated', 'employee.deleted',
            'approval.requested', 'approval.approved', 'approval.rejected'
        ];

        const events = Array.isArray(data.events) ? data.events : [data.events];
        const invalidEvents = events.filter(e => !validEvents.includes(e) && e !== '*');
        if (invalidEvents.length > 0) {
            return res.status(400).json({ error: `Invalid events: ${invalidEvents.join(', ')}` });
        }

        const db = getDb();
        const webhookDoc = {
            _id: new ObjectId(),
            companyId: isValidObjectId(data.companyId) ? new ObjectId(data.companyId) : data.companyId,
            url: data.url,
            events,
            secret: data.secret || generateSecret(),
            headers: data.headers || {},
            status: 'active',
            retryPolicy: {
                maxRetries: data.maxRetries || 3,
                retryDelay: data.retryDelay || 5000
            },
            createdAt: new Date(),
            lastUpdated: new Date()
        };

        await db.collection('webhooks').insertOne(webhookDoc);

        res.status(201).json({
            message: 'Webhook created successfully',
            _id: webhookDoc._id.toString(),
            secret: webhookDoc.secret
        });
    } catch (error) {
        console.error('Error creating webhook:', error);
        next(error);
    }
});

/**
 * PUT /api/webhooks/:webhook_id
 * Update webhook
 */
router.put('/:webhook_id', requireCompanyAccess, async (req, res, next) => {
    try {
        const { webhook_id } = req.params;
        const data = req.body;

        if (!isValidObjectId(webhook_id)) {
            return res.status(400).json({ error: 'Invalid webhook ID format' });
        }

        const updateFields = {};
        const allowedFields = ['url', 'events', 'headers', 'status', 'retryPolicy'];

        for (const field of allowedFields) {
            if (data[field] !== undefined) {
                updateFields[field] = data[field];
            }
        }

        if (Object.keys(updateFields).length === 0) {
            return res.status(400).json({ error: 'No fields to update' });
        }

        updateFields.lastUpdated = new Date();

        const db = getDb();
        const result = await db.collection('webhooks').updateOne(
            { _id: new ObjectId(webhook_id) },
            { $set: updateFields }
        );

        if (result.matchedCount === 0) {
            return res.status(404).json({ error: 'Webhook not found' });
        }

        res.json({ message: 'Webhook updated successfully' });
    } catch (error) {
        console.error('Error updating webhook:', error);
        next(error);
    }
});

/**
 * DELETE /api/webhooks/:webhook_id
 * Delete webhook
 */
router.delete('/:webhook_id', requireCompanyAccess, async (req, res, next) => {
    try {
        const { webhook_id } = req.params;

        if (!isValidObjectId(webhook_id)) {
            return res.status(400).json({ error: 'Invalid webhook ID format' });
        }

        const db = getDb();
        const result = await db.collection('webhooks').deleteOne({ _id: new ObjectId(webhook_id) });

        if (result.deletedCount === 0) {
            return res.status(404).json({ error: 'Webhook not found' });
        }

        res.json({ message: 'Webhook deleted successfully' });
    } catch (error) {
        console.error('Error deleting webhook:', error);
        next(error);
    }
});

/**
 * POST /api/webhooks/:webhook_id/test
 * Test webhook by sending a test event
 */
router.post('/:webhook_id/test', requireCompanyAccess, async (req, res, next) => {
    try {
        const { webhook_id } = req.params;

        if (!isValidObjectId(webhook_id)) {
            return res.status(400).json({ error: 'Invalid webhook ID format' });
        }

        const db = getDb();
        const webhook = await db.collection('webhooks').findOne({ _id: new ObjectId(webhook_id) });

        if (!webhook) {
            return res.status(404).json({ error: 'Webhook not found' });
        }

        // Send test event
        const testPayload = {
            event: 'test',
            timestamp: new Date().toISOString(),
            data: { message: 'This is a test webhook event' }
        };

        try {
            const response = await fetch(webhook.url, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                    'X-Webhook-Secret': webhook.secret,
                    ...webhook.headers
                },
                body: JSON.stringify(testPayload)
            });

            res.json({
                success: response.ok,
                statusCode: response.status,
                message: response.ok ? 'Test webhook sent successfully' : 'Webhook endpoint returned error'
            });
        } catch (fetchError) {
            res.json({
                success: false,
                error: fetchError.message,
                message: 'Failed to reach webhook endpoint'
            });
        }
    } catch (error) {
        console.error('Error testing webhook:', error);
        next(error);
    }
});

// Helper function
function generateSecret() {
    const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
    let secret = 'whsec_';
    for (let i = 0; i < 32; i++) {
        secret += chars.charAt(Math.floor(Math.random() * chars.length));
    }
    return secret;
}

module.exports = router;
